File: models/product.js
const mongoose=require("mongoose");
const ratingSchema = require("./ratings");

const productSchema=mongoose.Schema({
    name:{
        required:true,
        type:String,
        trim:true
    },
    description:{
        required:true,
        trim:true,
        type:String,
    },
    price:{
        required:true,
        type:Number,
        
    },
    quantity:{
        required:true,
        type:Number,
    },
    category:{
        required:true,
        type:String,
        trim:true
    },
    imageUrl:[{
        required:true,
        type:String,
    },
],
    ratings:[ratingSchema]

});

const Product=mongoose.model("Product",productSchema);
module.exports={Product,productSchema};

File: lib/model/product.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

import 'package:amazon_clone/model/rating.dart';

class Product {
  final String name;
  final String description;
  final double price;
  final int quantity;
  final String category;
  final List<String> imageUrl;
  final String? id;
  final List<Rating>? rating;

  Product(
      {required this.name,
      required this.description,
      required this.price,
      required this.quantity,
      required this.category,
      required this.imageUrl,
      this.id,
      this.rating});

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'name': name,
      'description': description,
      'price': price,
      'quantity': quantity,
      'category': category,
      'imageUrl': imageUrl,
      'id': id,
    };
  }

  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
      name: map['name'] ?? '',
      description: map['description'] ?? '',
      price: map['price']?.toDouble() ?? 0.0,
      quantity: map['quantity'] ?? 0,
      category: map['category'] ?? '',
      imageUrl: List<String>.from(map['imageUrl']),
      id: map['_id'] ?? '',
      rating: map['ratings'] != null
          ? List<Rating>.from(
              map['ratings']?.map(
                (x) => Rating.fromMap(x),
              ),
            )
          : null,
    );
  }

  String toJson() => json.encode(toMap());

  factory Product.fromJson(String source) =>
      Product.fromMap(json.decode(source) as Map<String, dynamic>);
}
